akka-http-core/src/main/scala/akka/http/impl/engine/http2/hpack/HeaderDecompression.scala

```diff
@@ -42,11 +47,34 @@ private[http2] object HeaderDecompression extends GraphStage[FlowShape[FrameEven
     // Receiving headers: waiting for CONTINUATION frame
 
     def parseAndEmit(streamId: Int, endStream: Boolean, payload: ByteString, prioInfo: Option[PriorityFrame]): Unit = {
-      val headers = new VectorBuilder[(String, String)]
+      val headers = new VectorBuilder[(String, AnyRef)]
       object Receiver extends HeaderListener {
-        def addHeader(name: Array[Byte], value: Array[Byte], sensitive: Boolean): Unit =
-          // TODO: optimization: use preallocated strings for well-known names, similar to what happens in HeaderParser
-          headers += new String(name, UTF8) -> new String(value, UTF8)
+        def addHeader(name: String, value: String, parsed: AnyRef, sensitive: Boolean): AnyRef = {
+          if (parsed ne null) {
+            headers += name -> parsed
+            parsed
+          } else {
+            import Http2HeaderParsing._
+            def handle(parsed: AnyRef): AnyRef = {
+              headers += name -> parsed
+              parsed
+            }
+
+            name match {
```

Comment:
We match by name instead of using a more dynamic structure (using the list of defined parsers in Http2HeaderParsing), because we expect the generated match being faster than a real HashMap (or similar). In fact, I looked at what Scala 2.13 generates for pure String matching. It basically uses a lookup switch on hashcodes, so that's basically a code-generated HashMap. Also this way, we avoid the overhead of dynamic dispatching the `parse()` invocations.